addons/binding/org.openhab.binding.tesla/ESH-INF/thing/models.xml

```diff
@@ -54,8 +54,8 @@
 			<channel id="headingestimation" typeId="headingestimation" />
 			<channel id="estimatedrange" typeId="estimatedrange" />
 <!-- 			<channel id="european" typeId="european" />
```

Comment:
Can you pease remove the commented out parts completely and their related channel-types as well.

Reply:
@martinvw No, I would not like to do that as it is a way to track the current state of the Tesla protocol.

Reply:
Imho git is the system we use to track the current state of things, if we want to preserve or view history the git log is the place to be. 

The current binding code in the openhab/openhab2-addons repository should reflect the current state of the API. Looking at the past 6 months it seems (from my perspective) that the API on Tesla's side is beginning to become less fluid and more mature.

It would be great if we could come to point were we could carefully inspect which fields are not available anymore and remove them from the code and the XML's. This makes the code more focused and easier to maintain. 

After that we should take a look at some refactoring to reduce the complexity and readability of the code. To ease refactoring we could also start by adding tests which prove that the new code behaves the same. As already promised before I'm willing to help in that refactoring, but I do not own a Tesla myself 😞 

Disclaimer: if there is really essential information to know / remember as developer of the binding we could just create a readme for developers where we can store relevant design decisions if there is no other way to make them clear / explicit in the code / (method) naming.

Reply:
@martinvw Tesla's API evolves all the time, sometimes there are major changes when for example a new vehicle/model is introduced, sometimes they are minor driven by the over-the-air updates of the software residing in the car. I expect it to continue to change, and I expect demands in that sense from other people as soon as Model 3 will be in the hands of consumers. I only have a Model S, so that is the only one I can code against.

With respect to the complexity of the code I can only say that it is what it is. Things would be more simple if Tesla did not make things complex, e.g. (1) this is a reverse engineering effort (2) they deviate from standards (in casu : their OAuth implementation is not following the standard) and (3) their servers are a complete blackbox with behaviour changing over time. (3) is causing the most trouble, as they (a) have massive growing pains and their servers can not keep up with the number of vehicles sold and in operation, causing (b) service outages, (c) service "resets" (d) forced log-outs, (e) blocking people that are hammering their servers and so forth. The back-end is not predictive, stable and all what you want. The code tries to work around all of these troubles. Therefore : If it ain't broken, don't fix it. I will leave any refactoring efforts to someone with *a lot* of time to debug and finetune. For me it works, therefore I won't touch it if not necessary.